### PR TITLE
dmsquash-live: add rd.live.join

### PR DESCRIPTION
**Pull request by udeved**
Support a union join on rootfs wiih another img.

This makes throw away layers possible, or starting different setups depending on rootfs and the join.img.

Example:
* the live bootloader features a base system and a desktop system entry, where the desktop fs resides in the join.img, while rootfs would be just a console base system.
* throw away live layer, the installer extracts the rootfs, while live session fs resides in the join.img, including the installer itself

Usage:

`rd.live.join=$name.img`

### CODE REVIEW

**Comment by haraldh** [comment]
@FGrose could you please review?

File: modules.d/90dmsquash-live/dmsquash-live-root.sh [hunk 1]
```diff
@@ -286,28 +287,49 @@ fi
...
+    SQUASHED_JOIN="/run/initramfs/live/${live_dir}/${join}"
+fi
 if [ -e "$SQUASHED" ]; then
     if [ -n "$live_ram" ]; then
         echo 'Copying live image to RAM...' > /dev/kmsg
         echo ' (this may take a minute)' > /dev/kmsg
         dd if=$SQUASHED of=/run/initramfs/squashed.img bs=512 2> /dev/null
         echo 'Done copying live image to RAM.' > /dev/kmsg
         SQUASHED="/run/initramfs/squashed.img"
+        if [ -n ${join} ];then
```
Comment (FGrose) [inline modules.d/90dmsquash-live/dmsquash-live-root.sh]:
```suggestion
        if [ -n ${join} ]; then
```
Did you intend to support the live_ram option?  If so, please verify that this new code path works.

File: modules.d/90dmsquash-live/dmsquash-live-root.sh [hunk 2]
```diff
@@ -286,28 +287,49 @@ fi
...
     SQUASHED_LOOPDEV=$( losetup -f )
     losetup -r $SQUASHED_LOOPDEV $SQUASHED
     mkdir -m 0755 -p /run/initramfs/squashfs
     mount -n -t squashfs -o ro $SQUASHED_LOOPDEV /run/initramfs/squashfs
 
+    if [ -n ${join} ]; then
+        SQUASHED_JOIN_LOOPDEV=$( losetup -f )
+        losetup -r $SQUASHED_JOIN_LOOPDEV $SQUASHED_JOIN
+        mkdir -m 0755 -p /run/initramfs/joinfs
+        mount -n -t squashfs -o ro $SQUASHED_LOOPDEV /run/initramfs/joinfs
```
Comment (FGrose) [inline modules.d/90dmsquash-live/dmsquash-live-root.sh]:
```suggestion
        mount -n -t squashfs -o ro $SQUASHED_JOIN_LOOPDEV /run/initramfs/joinfs
```
Did you intend to support an unflat join.img?  I would recommend not to as you require an OverlayFS for the union join.  (You don't use the /run/initramfs/joinfs mountpoint anyway.  This section is superfluous.)

File: modules.d/90dmsquash-live/dmsquash-live-root.sh [hunk 3]
```diff
@@ -286,28 +287,49 @@ fi
...
+    fi
+
     if [ -d /run/initramfs/squashfs/LiveOS ]; then
         if [ -f /run/initramfs/squashfs/LiveOS/rootfs.img ]; then
             FSIMG="/run/initramfs/squashfs/LiveOS/rootfs.img"
         elif [ -f /run/initramfs/squashfs/LiveOS/ext3fs.img ]; then
             FSIMG="/run/initramfs/squashfs/LiveOS/ext3fs.img"
         fi
+        if [ -f /run/initramfs/squashfs/LiveOS/${join} ]; then
```
Comment (FGrose) [inline modules.d/90dmsquash-live/dmsquash-live-root.sh]: This is looking for the join.img inside the squashfs.img.  Not correct.

File: modules.d/90dmsquash-live/dmsquash-live-root.sh [hunk 4]
```diff
@@ -323,12 +345,22 @@ else
     elif [ -e /run/initramfs/live/${live_dir}/ext3fs.img ]; then
         FSIMG="/run/initramfs/live/${live_dir}/ext3fs.img"
     fi
+    if [ -e /run/initramfs/live/${live_dir}/${join} ]; then
```
Comment (FGrose) [inline modules.d/90dmsquash-live/dmsquash-live-root.sh]: Did you intend to support an uncompressed join.img?  If so, please verify that this new code path works.

File: modules.d/90dmsquash-live/dmsquash-live-root.sh [hunk 5]
```diff
@@ -343,6 +375,15 @@ if [ -n "$FSIMG" ]; then
             unpack_archive $FSIMG /run/initramfs/fsimg/
         fi
         FSIMG=/run/initramfs/fsimg/rootfs.img
+
+        if [ -n ${join} ]; then
```
Comment (FGrose) [inline modules.d/90dmsquash-live/dmsquash-live-root.sh]: Did you intend to support the 'writeable_fsimg' option?  If so, please verify that this new code path works.

File: modules.d/90dmsquash-live/dmsquash-live-root.sh [hunk 6]
```diff
@@ -358,12 +399,23 @@ if [ -n "$FSIMG" ]; then
     fi
     if [ "$FSIMG" = "$SQUASHED" ]; then
         BASE_LOOPDEV=$SQUASHED_LOOPDEV
+        if [ -n ${join} ]; then
```
Comment (FGrose) [inline modules.d/90dmsquash-live/dmsquash-live-root.sh]: Device-mapper based configurations will not support a union join of rootfs content with joinfs.  This code is superfluous, no?

File: modules.d/90dmsquash-live/dmsquash-live-root.sh [hunk 7]
```diff
@@ -390,12 +442,26 @@ if [ -n "$overlayfs" ]; then
...
+            ovlfs=lowerdir=/run/rootfsjoin:/run/rootfsbase
+        else
+            ovlfs=lowerdir=/run/rootfsbase
+        fi
     fi
     mount -r $FSIMG /run/rootfsbase
+    if [ -n ${join} ]; then
+        mount -r $FSIMG_JOIN /run/rootfsjoin
+    fi
     if [ -z "$DRACUT_SYSTEMD" ]; then
```
Comment (FGrose) [inline modules.d/90dmsquash-live/dmsquash-live-root.sh]: In order to support DRACUT_SYSTEMD systems, the ovlfs variable in dmsquash-generator.sh must be similarly modified.

**Comment by udeved** [comment]
@FGrose 

I am sorry for the late reply, as soon as I have enough time at hand, I will look more closely at it.
We still haven't switched to dracut, and this rough patch did the job to test our install media boot with dracut.
Tested was a flat squash image and a squashed ext image.

**Comment by danimo** [comment]
@udeved any update?

**Pull request closed without merging**